avoid intermediate map allocations in multi-arg mapreduce

mbauman opened the pull request:
Now that the mapreduce infrastructure (mostly) supports broadcasted objects, we can use a lazy Broadcasted on array-likes instead of using an intermediate `map`. This exercises the internals more thoroughly and identifies a few more places where we need to use `::AbstractArrayOrBroadcasted` and requires an offset axis bugfix.

Fixes #38558 

<details><summary>local benchmarks</summary>

```julia-repl
julia> using BenchmarkTools

julia> function f(x, y)
           return @inbounds mapreduce(==,+,x, y)
       end
f (generic function with 1 method)

julia> function f2(x, y)
               total=0
               @inbounds for i in 1:length(x)
                       total += x[i]==y[i]
               end
               return total
       end
f2 (generic function with 1 method)

julia> x = randn(10240); y = similar(x);

julia> @btime f2(x,y)
  1.988 μs (0 allocations: 0 bytes)
0

julia> @btime f(x,y)
  2.023 μs (0 allocations: 0 bytes)
0
```
</details>

Alternative to:  #41001 (cc @mcabbott)

Fixes #53417

<details><summary>more benchmarking from 53417</summary>

```julia-repl
julia> x = randn((512,512));

julia> y = randn((512, 512));

julia> g(x) = x^2
g (generic function with 1 method)

julia> @time mapreduce(g,+,x)
  0.031427 seconds (90.73 k allocations: 4.607 MiB, 24.14% gc time, 99.23% compilation time)
262233.95257231314

julia> @time mapreduce(g,+,x)
  0.000064 seconds (1 allocation: 16 bytes)
262233.95257231314

julia> f(x,y) = x * y
f (generic function with 1 method)

julia> mapreduce(f, +, x, y);

julia> @time mapreduce(f, +, x, y);
  0.000746 seconds (3 allocations: 80 bytes)
```

</details>

mbauman committed:
avoid intermediate map allocations in multi-arg mapreduce

Now that the mapreduce infrastructure (mostly) supports broadcasted objects, we can use a lazy Broadcasted on array-likes instead of using an intermediate `map`. This exercises the internals more thoroughly and identifies a few more places where we need to use `::AbstractArrayOrBroadcasted` and requires an offset axis bugfix.

N5N3 commented on base/broadcast.jl:
```diff
@@ -578,7 +578,7 @@ an `Int`.
 Base.@propagate_inbounds newindex(arg, I::CartesianIndex) = to_index(_newindex(axes(arg), I.I))
 Base.@propagate_inbounds newindex(arg, I::Integer) = to_index(_newindex(axes(arg), (I,)))
-Base.@propagate_inbounds _newindex(ax::Tuple, I::Tuple) = (ifelse(length(ax[1]) == 1, ax[1][1], I[1]), _newindex(tail(ax), tail(I))...)
+Base.@propagate_inbounds _newindex(ax::Tuple, I::Tuple) = (ifelse(length(ax[1]) == 1, ax[1][begin], I[1]), _newindex(tail(ax), tail(I))...)
```
Since we are at this, would you mind take a look at #43414?
(And maybe #43736?)

mbauman replied:
Oh, those are much better wholistic fixes!  Yes, let's get those both in.

N5N3 reviewed (commented):
The change on mapreduce is a broken one as `map(f, c...)` has no length constraint if `ndims` of inputs are `1`
```julia
julia> mapreduce(+, +, 1:3, 1:2) == mapreduce(+, +, 1:2, 1:2)
true
```
But the rest seems good though, see also #41054.

mbauman commented:
> ```julia
> mapreduce(+, +, 1:3, 1:2) == mapreduce(+, +, 1:2, 1:2)
> ```

😭

Man, this is why we can't have nice things. Pretty wildly, this behavior doesn't look to be tested.  I could've sworn we had an absurdly long megathread on this (mis)feature, but I can't find it at the moment.

mcabbott commented:
I think I added the documentation at some point to match the reality... which IIRC was added basically by accident in 1.5, and had no tests for ages. On 1.0 `map` gave an error instead:
```julia
julia> VERSION
v"1.0.5"

julia>  map(+, 1:2, 1:3)
ERROR: DimensionMismatch("dimensions must match")
```

mbauman commented:
Yeah, but the iterator version has done it "forever":

```julia-repl
julia> VERSION
v"1.0.5"

julia> map(+, Iterators.drop(1:5, 1), Iterators.drop(1:5, 2))
3-element Array{Int64,1}:
 5
 7
 9
```

It's somewhat related to #46707, but I still can't find the discussion I was thinking of.

mbauman committed:
restore support for args of different dimensionalities

this now matches the documentation, but still is not *exactly* the previous behavior

mbauman commented:
So this is still slightly more restrictive than what we had before. Let's see what Nanosoldier says:

@nanosoldier runtests()

KristofferC commented:
@nanosoldier `runtests()`

N5N3 commented on base/reducedim.jl:
```diff
@@ -87,7 +94,7 @@ end
 # initialization when computing minima and maxima requires a little care
 for (f1, f2, initval, typeextreme) in ((:min, :max, :Inf, :typemax), (:max, :min, :(-Inf), :typemin))
-    @eval function reducedim_init(f, op::typeof($f1), A::AbstractArray, region)
+    @eval function reducedim_init(f, op::typeof($f1), A::AbstractArrayOrBroadcasted, region)
```
This change seems incomplete as `view` in L106 has no `Broadcasted` support.
we might need
```julia
reduced_view(A::AbstractArray, ri) = view(A, ri...)
reduced_view(bc::Broadcasted, ri) = Broadcasted(bc.style, bc.f, bc.args, ri)
```

nanosoldier commented:
The package evaluation job [you requested](https://github.com/JuliaLang/julia/pull/55301#issuecomment-2258388388) has completed - possible new issues were detected.
The [**full report**](https://s3.amazonaws.com/julialang-reports/nanosoldier/pkgeval/by_hash/07f320a_vs_4dfce5d/report.html) is available.

mbauman commented:
I'll put this on hold for now — #55318 will unblock the hard part here.
